### PR TITLE
chore: implement classnames

### PR DESCRIPTION
**Pull request by mobot11**
This pr replaces all instances where classes were being computed in `className` and moves the logic into the `classnames` module. This is a huge pr, so I'd like to get multiple eyes on it. Fixes https://github.com/SAP/fundamental-react/issues/16.

### CODE REVIEW

File: src/Alert/Alert.js [hunk 1]
```diff
@@ -9,55 +10,66 @@ export class Alert extends Component {
...
-              )}
-          </div>
-      );
-  }
+    render() {
+        const {
+            type,
+            link,
+            linkText,
+            dismissible,
```
Comment (bcullman) [inline src/Alert/Alert.js]: you changed `dismissable` to `dismissible`

that is a breaking change and while we should allow for breaking changes while we are 0.X.X, we should not intermix that work with this PR.

Comment (mobot11) [reply src/Alert/Alert.js]: Sounds good. I saw that they were different and assumed they should match. I'll open an issue to fix instead.

Comment (mobot11) [reply src/Alert/Alert.js]: https://github.com/SAP/fundamental-react/issues/234

File: src/Badge/Badge.js [hunk 2]
```diff
@@ -39,8 +78,16 @@ Status.propTypes = {
...
+        'fd-counter',
+        {
+            'fd-counter--notification': notification
+        },
+        className
+    );
+
     return (
-        <span aria-label='Unread count' className={`fd-counter${notification ? ' fd-counter--notification' : ''}${className ? ' ' + className : ''}`}
+        <span aria-label='Unread count' className={counterClasses}
```
Comment (greg-a-smith) [inline src/Badge/Badge.js]: The `props` spread should come first or `aria-label` could be overridden by the consumer.

File: src/Button/Button.js [hunk 3]
```diff
@@ -15,11 +16,33 @@ export const Button = ({
...
+            'fd-button--positive': type === 'positive',
+            'fd-button--negative': type === 'negative',
+            'fd-button--medium': type === 'medium',
+            'fd-dropdown__control': dropdown,
+            'fd-button--compact': compact,
+            [`sap-icon--${glyph}`]: glyph,
+            'fd-global-nav__btn': navbar,
+            'is-selected': selected,
+            'is-disabled': disabled
+        },
```
Comment (greg-a-smith) [inline src/Button/Button.js]: Can the extensible pattern be used here for `option` and `type`?

**Comment by greg-a-smith** [comment]
@mobot11 Additionally, I would change the PR title prefix to `chore:` since we are not functionally changing anything -- just simply making use of an external library to clean the code up.

File: src/Icon/Icon.js [hunk 4]
```diff
@@ -17,5 +27,5 @@ Icon.propTypes = {
     glyph: PropTypes.string.isRequired,
     className: PropTypes.string,
     clickHandler: PropTypes.func,
-    size: PropTypes.string
+    size: PropTypes.oneOf(['', 'xxs', 'xs', 's', 'm', 'l', 'xl', 'xxl' ])
```
[the remark this answers is not in the document]
Comment (greg-a-smith) [reply src/Icon/Icon.js]: Actually, after I think about this some more, I think this change should come out of this PR.  It's probably a good change, but not for this PR.

File: src/Identifier/Identifier.js [hunk 5]
```diff
@@ -27,5 +37,5 @@ Identifier.propTypes = {
     glyph: PropTypes.string,
     label: PropTypes.string,
     modifier: PropTypes.oneOf(['', 'circle', 'transparent']),
-    size: PropTypes.string
+    size: PropTypes.oneOf(['', 'xxs', 'xs', 's', 'm', 'l', 'xl', 'xxl' ])
```
[the remark this answers is not in the document]
Comment (greg-a-smith) [reply src/Identifier/Identifier.js]: Actually, after I think about this some more, I think this change should come out of this PR.  It's probably a good change, but not for this PR.

File: src/Icon/Icon.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (bcullman) [reply src/Icon/Icon.js]: agreed. not a classname PR

in nui widgets, we maintain a master list of const for in a file called styleConfig.js. for example:

```
export const MESSAGEBAR_TYPES = [
    'success',
    'info',
    'warning',
    'error'
];

export const MESSAGEBOX_TYPES = [
    'confirm',
    'error',
    'info',
    'success',
    'warning'
];

export const BUTTON_TYPES = [
    'default',
    'primary',
    'secondary',
    'create',
    'success',
    'info',
    'warning',
    'danger',
    'link'
];
```

The name of that file is debate-able (as is if they should be centralized at all in a single file) but I do like using a const for this.

but again to @greg-a-smith point - this is not the PR for this change.

[the remark this answers is not in the document]
Comment (mobot11) [reply src/Icon/Icon.js]: I'll create an issue for this.

[the remark this answers is not in the document]
Comment (mobot11) [reply src/Icon/Icon.js]: https://github.com/SAP/fundamental-react/issues/269

**Pull request merged**